core/execution/engine_checkpoint.go

```diff
@@ -0,0 +1,47 @@
+package execution
+
+import (
+	"context"
+	"sort"
+
+	"code.vegaprotocol.io/vega/core/types"
+	"code.vegaprotocol.io/vega/libs/proto"
+	checkpoint "code.vegaprotocol.io/vega/protos/vega/checkpoint/v1"
+)
+
+func (e *Engine) Name() types.CheckpointName {
+	return types.ExecutionCheckpoint
+}
+
+func (e *Engine) Checkpoint() ([]byte, error) {
+	for id, mkt := range e.markets {
+		state := mkt.GetCPState()
+		e.marketCPStates[id] = state
+	}
+	data := make([]*types.CPMarketState, 0, len(e.marketCPStates))
+	// @TODO prune the marketCPStates
+	for _, s := range e.marketCPStates {
+		data = append(data, s)
+	}
+	sort.SliceStable(data, func(i, j int) bool {
+		return data[i].ID < data[j].ID
+	})
+	wrapped := types.ExecutionState{
+		Data: data,
+	}
+	cpData, err := proto.Marshal(wrapped.IntoProto())
+	return cpData, err
+}
+
+func (e *Engine) Load(ctx context.Context, data []byte) error {
+	wrapper := checkpoint.ExecutionState{}
+	if err := proto.Unmarshal(data, &wrapper); err != nil {
+		return err
+	}
+	e.marketCPStates = make(map[string]*types.CPMarketState, len(wrapper.Data))
+	// for now, restore all pending markets as though their state is valid for the full TTL window
+	for _, mcp := range wrapper.Data {
+		e.marketCPStates[mcp.Id] = types.NewMarketStateFromProto(mcp)
```

Comment:
anything to do with the maps of isSuccessor, etc that were added as well?

Reply:
Yes, this is needed in case a successor market is proposed during the successor window (grace period), after the market is settled. The settled market will be removed from the normal `e.markets` and `marketsCpy` fields as usual, but we keep settled markets around in case there is a successor market on the way. We also need to be able to create a successor market correctly from a checkpoint restore (which uses proposals, and therefore doesn't contain the necessary ELS data).

TL;DR: yes, it's related, and both are necessary